### ISSUE TITLE
Introduce event hook for auth/accessToken

### ISSUE THREAD

**Issue by kbrabrand** [imbo/imbo#326]
The event subscription lists in `EventListener\Authenticate` and `EventListener\AccessToken` specifies what endpoint requires authentication. This work well for core resources, but for plugins (for instance the metadata search plugin) it's a bit cumbersome to work with.

If we introduced listeners called `authenticate` and `accessToken` respectively we would have a way of triggering the authentication from third party modules as well.

Unless I'm missing something and this functionality already exists..

Any thoughts on this?

**Comment by rexxars** [comment]
Couldn't you just trigger `authenticate()` manually, from the resource?

Perhaps a bit tight coupling... There would be no way to disable it, etc.

**Comment by rexxars** [comment]
I guess we could just add `auth.authenticate` and `auth.accesstoken` to the list of events the respective listeners already subscribe to. Then it would be as easy as triggering a new event. I think  @christeredvartsen should weigh in on this though.

**Comment by kbrabrand** [comment]
Not sure how I would "trigger `authenticate()` it manually".. Right now the cleanest way seem to be extending the AccessToken class inside the plugin and specifying the event subscriptions in there. Like this;

```php
<?php

namespace Imbo\MetadataSearch\EventListener;

use Imbo\EventListener\AccessToken as ImboAccessToken;

/**
 * Access token event listener used for authenticating metadata
 * endpoints
 */
class AccessToken extends ImboAccessToken {
    public static function getSubscribedEvents() {
        return [
            'metadata.search' => ['checkAccessToken' => 100]
        ];
    }
}
```

but it would be nice to simply trigger `auth.authenticate` or `auth.accesstoken` instead..

**Comment by rexxars** [comment]
The resource or eventlistener gets an event as argument. Use it to get the instance of the event manager:

`$event->getManager()->trigger('auth.authenticate');`

**Comment by kbrabrand** [comment]
Yes but no. I tried that before opening ticket (and again right now), and it works fine if the authenticate listener is subscribing to that event name (`auth.authenticate`), but unless it does, nothing happens.

**Comment by rexxars** [comment]
Obviously, which was my suggestion - adding these two event to the core listeners.

I just want a second opinion on it first.

**Comment by kbrabrand** [comment]
Then we're trying to say the same thing :smile:

**Comment by rexxars** [comment]
Implemented in  #329
